lib/jormungandr/src/Cardano/Wallet/Binary/Jormungandr.hs

```diff
@@ -8,5 +11,173 @@
+    ( Base (Base16), convertFromBase )
+import Data.ByteString
+    ( ByteString )
+import Data.Word
+    ( Word16, Word32 )
+
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as BL
+
+
+-- | Messages is what the block body consists of.
+--
+-- Every message is prefixed with a message header.
+--
+--  Following, as closely as possible:
+-- https://github.com/input-output-hk/rust-cardano/blob/e0616f13bebd6b908320bddb1c1502dea0d3305a/chain-impl-mockchain/src/message/mod.rs#L22-L29
+data Message
+    = Initial [ConfigParam]
+    | OldUtxoDeclaration TODO
+    | Transaction TODO
+    | Certificate TODO
+    | UpdateProposal SignedUpdateProposal
+    | UpdateVote SignedVote
+    | UnimplementedMessage Int -- For development. Remove later.
+    deriving Show
+
+
+data BlockHeader = BlockHeader
+    { version :: Word16
+    , contentSize :: Word32
+    , slot :: SlotId
+    , chainLength :: Word32
+    , contentHash :: ByteString
+    , parentHeaderHash :: Maybe ByteString
+    } deriving (Show, Eq)
+
+data Block = Block BlockHeader [Message]
+    deriving Show
+
+
+data SignedUpdateProposal = SignedUpdateProposal
+    deriving Show
+data TODO = TODO
+    deriving Show
+data SignedVote = SignedVote
+    deriving Show
+data ConfigParam = ConfigParam
+    deriving Show
+
+{-# ANN getBlockHeader ("HLint: ignore Use <$>" :: String) #-}
+getBlockHeader :: Get BlockHeader
+getBlockHeader =  (fromIntegral <$> getWord16be) >>= \s -> isolate s $ do
+    version <- getWord16be
+    sizeOfContent <- getWord32be
+
+    slotEpoch <- fromIntegral <$> getWord32be
+    slotId <- fromIntegral <$> getWord32be
+
+    chainLength <- getWord32be
+
+    contentHash <- getByteString 32 -- or 256 bits
+
+    parentHeaderHash <- getParentHeaderHash
+
+    -- TODO: Handle special case for BFT
+    -- TODO: Handle special case for Praos/Genesis
+
+    return $
+        BlockHeader
+            version
+            sizeOfContent
+            (SlotId slotId slotEpoch)
+            chainLength
+            contentHash
+            parentHeaderHash
+
+getBlock :: Get Block
+getBlock = do
+    header <- getBlockHeader
+    msgs <- isolate (fromIntegral $ contentSize header)
+        $ whileM (not <$> isEmpty) getBlockContent
+    return $ Block header msgs
+
+getBlockContent :: Get Message
+getBlockContent = do
+    size <- getWord16be
+    contentType <- fromIntegral <$> getWord8
+
+    let msgSize = fromIntegral size - 1
+
+    let unimpl = skip msgSize >> return (UnimplementedMessage contentType)
+
+    isolate msgSize $ case contentType of
+        0 -> unimpl
+        1 -> unimpl
+        2 -> unimpl
+        3 -> unimpl
+        4 -> unimpl
+        5 -> unimpl
+        other -> fail $ "Unexpected content type tag " ++ show other
+
+getParentHeaderHash :: Get (Maybe ByteString)
+getParentHeaderHash = getByteString 32 >>= \case
+    a | a == BS.pack (replicate 32 0)
+        -> return Nothing
+      | otherwise
+        -> return $ Just a
```

Comment:
It seems that there's a parent header hash, isn't it? Since most blocks except the first one would have a parent header hash that is significant, we're probably better off having a `ByteString` (or `Hash "parent"`, or whatnot) instead of a `Maybe ...` and handle that first case as an edge-case somewhere.

Reply:
Now have
```haskell
    , contentHash :: Hash "content"
    , parentHeaderHash :: Hash "parentHeader"
```

Slight hesitation with "parentHeader" vs "header", but for clarity, definitely "parentHeader".